diff --git a/dmenu.c b/dmenu.c
index 2f2d331..2740fa9 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -28,7 +28,8 @@
 #define LENGTH(X)             (sizeof X / sizeof X[0])
 #define TEXTW(X)              (drw_fontset_getwidth(drw, (X)) + lrpad)
 // >>>>>>>>>>>>>>>>>>>> numbers
-// ==================== numbers
+#define NUMBERSMAXDIGITS      100
+#define NUMBERSBUFSIZE        (NUMBERSMAXDIGITS * 2) + 1
 // <<<<<<<<<<<<<<<<<<<< numbers
 
 /* enums */
@@ -65,7 +66,7 @@ static int mon = -1, screen;
 // ==================== xyw
 // <<<<<<<<<<<<<<<<<<<< xyw
 // >>>>>>>>>>>>>>>>>>>> numbers
-// ==================== numbers
+static char numbers[NUMBERSBUFSIZE] = "";
 // <<<<<<<<<<<<<<<<<<<< numbers
 
 static Atom clip, utf8;
@@ -170,7 +171,20 @@ drawitem(struct item *item, int x, int y, int w)
 }
 
 // >>>>>>>>>>>>>>>>>>>> numbers
-// ==================== numbers
+static void
+recalculatenumbers()
+{
+  unsigned int numer = 0, denom = 0;
+  struct item *item;
+  if (matchend) {
+    numer++;
+  for (item = matchend; item && item->left; item = item->left)
+    numer++;
+  }
+  for (item = items; item && item->text; item++)
+    denom++;
+  snprintf(numbers, NUMBERSBUFSIZE, "%d/%d", numer, denom);
+}
 // <<<<<<<<<<<<<<<<<<<< numbers
 
 static void
@@ -199,7 +213,7 @@ drawmenu(void)
 	}
 
   // >>>>>>>>>>>>>>>>>>>> numbers
-  // ==================== numbers
+  recalculatenumbers();
   // <<<<<<<<<<<<<<<<<<<< numbers
 	if (lines > 0) {
     // >>>>>>>>>>>>>>>>>>>> grid
@@ -219,20 +233,19 @@ drawmenu(void)
 		x += w;
 		for (item = curr; item != next; item = item->right)
       // >>>>>>>>>>>>>>>>>>>> numbers
-      // ==================== numbers
-      x = drawitem(item, x, 0, MIN(TEXTW(item->text), mw - x - TEXTW(">")));
+      x = drawitem(item, x, 0, MIN(TEXTW(item->text), mw - x - TEXTW(">") - TEXTW(numbers)));
       // <<<<<<<<<<<<<<<<<<<< numbers
 		if (next) {
 			w = TEXTW(">");
 			drw_setscheme(drw, scheme[SchemeNorm]);
       // >>>>>>>>>>>>>>>>>>>> numbers
-      // ==================== numbers
-      drw_text(drw, mw - w, 0, w, bh, lrpad / 2, ">", 0);
+      drw_text(drw, mw - w - TEXTW(numbers), 0, w, bh, lrpad / 2, ">", 0);
       // <<<<<<<<<<<<<<<<<<<< numbers
 		}
 	}
   // >>>>>>>>>>>>>>>>>>>> numbers
-  // ==================== numbers
+  drw_setscheme(drw, scheme[SchemeNorm]);
+  drw_text(drw, mw - TEXTW(numbers), 0, TEXTW(numbers), bh, lrpad / 2, numbers, 0);
   // <<<<<<<<<<<<<<<<<<<< numbers
 	drw_map(drw, win, 0, 0, mw, mh);
 }
